x/stakeibc/keeper/icqcallbacks_validator_exchange_rate.go

```diff
@@ -76,18 +65,94 @@ func ValidatorExchangeRateCallback(k Keeper, ctx sdk.Context, args []byte, query
 	//     exchange_rate = num_tokens / num_shares
 	//  We can use `validator.TokensFromShares`, plug in 1.0 for the number of shares,
 	//    and the returned number of tokens will be equal to the internal exchange rate
-	validator.InternalExchangeRate = &types.ValidatorExchangeRate{
-		InternalTokensToSharesRate: queriedValidator.TokensFromShares(sdk.NewDec(1.0)),
-		EpochNumber:                strideEpochTracker.GetEpochNumber(),
-	}
+	currentSharesToTokensRate := queriedValidator.TokensFromShares(sdk.NewDec(1.0))
+	validator.InternalSharesToTokensRate = currentSharesToTokensRate
 	hostZone.Validators[valIndex] = &validator
 	k.SetHostZone(ctx, hostZone)
 
-	k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator, "Validator Internal Exchange Rate: %v",
-		validator.InternalExchangeRate.InternalTokensToSharesRate))
+	// Check if the validator was slashed by comparing the exchange rate from the query
+	// with the preivously stored exchange rate
+	validatorWasSlashed := false
+	if previousSharesToTokensRate.IsNil() || previousSharesToTokensRate.IsZero() {
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Previous Validator Exchange Rate Unknown"))
+
+	} else {
+		validatorWasSlashed = !previousSharesToTokensRate.Equal(currentSharesToTokensRate)
+
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Previous Validator Exchange Rate: %v, Current Validator Exchange Rate: %v",
+			previousSharesToTokensRate, currentSharesToTokensRate))
+	}
+
+	// Determine if we're in a callback for the LSMLiquidStake by checking if the callback data is non-empty
+	// If this query was triggered manually, the callback data will be empty
+	inLSMLiquidStakeCallback := len(query.CallbackData) != 0
+
+	// If the validator was not slashed, and we're NOT in the callback of an LSM Liquid stake,
+	// we can stop here - there's no need to query for the delegator shares
+	if !validatorWasSlashed && !inLSMLiquidStakeCallback {
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(chainId, ICQCallbackID_Validator,
+			"Validator was not slashed"))
+		return nil
+	}
+
+	// If we are in an LSMLiquidStake callback, unmarshal the callback data
+	var lsmLiquidStake types.LSMLiquidStake
+	var lsmTokenDeposit recordstypes.LSMTokenDeposit
+	if inLSMLiquidStakeCallback {
+		var callbackData types.ValidatorExchangeRateQueryCallback
+		if err := proto.Unmarshal(query.CallbackData, &callbackData); err != nil {
+			return errorsmod.Wrapf(err, "unable to unmarshal validator exchange rate callback data")
+		}
+		lsmLiquidStake = *callbackData.LsmLiquidStake
+		lsmTokenDeposit = *lsmLiquidStake.Deposit
+	}
+
+	// If the validator was not slashed, and we're in the callback of an LSM liquid stake,
+	// finish the transaction to mint the user their stTokens
+	if !validatorWasSlashed && inLSMLiquidStakeCallback {
```

Comment:
agreed it is a bit tough to follow. Can you clarify your suggestion? You're saying a separate function for just the LSM branch?

Reply:
I was saying two different callbacks named different things based on whether we are doing the normal flow, or the async flow.  The comments are good and clear, I just was finding the overall structure of these cases hard to grasp because we reach the same point in the function in different control flows.  I don't know if my idea really simplifies it much